Explicit tally moment order strings in statepoint files

wbinventor opened the pull request:
This pull request modifies OpenMC's ``statepoint.*`` files to include explicit strings moment orders with for tallies with scores such as ``flux-PN``, ``scatter-YN``, etc at the request of @bforget. The current release version of openmc prints an integer array of the moment orders for each score. The issue with this is it is not transparent to the user which tally results correspond to which moment based on looking at the ``statepoint.h5`` files. For example, if a user assigns a ``scatter-P3`` score to a tally, this implicitly defines the four scores: ``scatter-0``, ``scatter-1``, ``scatter-2`` and ``scatter-3``. However, the moment order array in the ``statepoint.*`` file appears as ``[3,3,3,3]`` for each of these scores. This problem is exacerbated for spherical harmonic moments.

This PR explicitly writes strings for each moment to the statepoint file. In the example mentioned above, the new ``statepoint.*`` files will now have strings under the ``moment`` group (in HDF5) indexed by score with strings ```"P0"``, ``"P1"``, ``"P2"``, and ``"P3"``, respectively. Likewise, for spherical harmonics the strings will appear as ``"Y0,0"``, ``"Y1,2"``, etc. These strings are also now properly parsed with the ``StatePoint`` Python class for use by the user. Note that this functionality is an extension of the code implemented by @nelsonag for the ``tallies.out`` files in ``output.F90``, though modified somewhat to work with binary or HDF ``statepoint`` files.

wbinventor commented:
Just curious - is there any hangup with this PR that needs to be addressed @nelsonag @paulromano @bhermanmit?

paulromano commented:
I was waiting for the 0.6.2 release to be complete. Now that it is, we'll be working to clear out the backlog of PRs to be included in a 0.7 release.

paulromano commented on src/utils/statepoint.py:
```diff
@@ -290,15 +290,28 @@ def _read_metadata(self):
+              moment.replace(" ", "")
+
+              # Add the moment to the tally's list
+              t.moments.append(moment)
```
Four spaces for indentation on this block

paulromano commented on src/state_point.F90:
```diff
@@ -214,18 +218,59 @@ subroutine write_state_point()
+              end do
+            end do
+          case default
+            moment_name = '0'
```
Might it be better to have an empty string rather than '0'?

paulromano commented on src/state_point.F90:
```diff
@@ -214,18 +218,59 @@ subroutine write_state_point()
+            moment_name = 'P' // to_str(t % moment_order(k))
+            call sp % write_data(moment_name, "order", &
+                 group="tallies/tally" // trim(to_str(i)) // &
+                 "/moments/" // trim(to_str(k)))
```
I understand the motivation to have a separate group for each moment string (output_interface API doesn't provide a way to write an array of strings). It might be better though to just have a separate dataset for each moment string in the tallies/tallyN/moments group, i.e. moments/order1, moments/order2, etc., until output_interface is improved.

wbinventor replied:
Could you refresh my memory on how I would change this function call to do that? I'm not really sure I see what the benefit is in making this change - it seems to me that including each one as a separate group with the order index explicitly defined is more transparent to the user.

paulromano replied:
It would look like
``` Fortran
call sp % write_data(moment_name, "order" // trim(to_str(k)), &
    group="tallies/tally" // trim(to_str(i)) // "/moments")
```

In my mind, they are equivalent, except using datasets only means you avoid creating N groups which saves a little bit of space and time. This is just a nitpick, so if you feel strongly we can leave as is. Eventually we should move to having an array of strings but we needn't do it as part of this PR.

paulromano replied:
I think it would also remove the need to do the open_group/close_group inside the `#ifdef HDF5` above.

wbinventor replied:
Thanks for hint - I just made the change for both writing/loading data in ``state_point.F90``, as well as the appropriate modification to ``statepoint.py``.

paulromano commented:
Two requests:

1. Add documentation on the new statepoint binary format (see `docs/source/devguide/statepoint.rst`)
2. Merge develop in so we can run the test suite. 

After that I think this should be good to go.

wbinventor commented:
Thanks for the reveiew @paulromano. I've made all of your requested changes, but have not addressed your comment about the HDF5 group vs. dataset since I'll have to spend some time re-learning how to use the statepoint interface to do this.

wbinventor commented:
@paulromano - I believe that all of your concerns have been addressed. The test suite runs fine on my machine. Let me know if you see anything else awry.

paulromano commented on src/state_point.F90:
```diff
@@ -214,18 +218,55 @@ subroutine write_state_point()
+#ifdef HDF5
+        call sp % open_group("tallies/tally" // trim(to_str(i)) // "/moments")
+        call sp % close_group()
+#endif
```
Can this #ifdef block be removed now? If I'm not mistaken, the call to `sp % write_data` will automatically create the group if it doesn't exist.

wbinventor replied:
Thanks for the catch - I just removed it.

paulromano commented:
Great. I'll run the test suite tonight and if everything is well, I'll merge it in. Thanks!

paulromano commented:
Tests [pass](http://openmc.mit.edu/cdash/index.php?project=OpenMC&date=2015-02-12&filtercount=1&showfilters=1&field1=buildname/string&compare1=63&value1=scatt-moment). There was a build failure on the PETSc configurations but that is unrelated to this PR.
